src/types/list.rs

```diff
@@ -253,51 +515,99 @@ impl PyList {
     ///
     /// This is equivalent to the Python expression `self.index(value)`.
     #[inline]
-    pub fn index<V>(&self, value: V) -> PyResult<usize>
+    fn index<V>(&self, value: V) -> PyResult<usize>
     where
         V: ToPyObject,
     {
         self.as_sequence().index(value)
     }
 
     /// Returns an iterator over this list's items.
-    pub fn iter(&self) -> PyListIterator<'_> {
-        PyListIterator {
-            list: self,
-            index: 0,
-            length: self.len(),
-        }
+    fn iter(&self) -> PyListIterator2<'py> {
+        PyListIterator2::new(self.clone())
     }
 
     /// Sorts the list in-place. Equivalent to the Python expression `l.sort()`.
-    pub fn sort(&self) -> PyResult<()> {
+    fn sort(&self) -> PyResult<()> {
         err::error_on_minusone(self.py(), unsafe { ffi::PyList_Sort(self.as_ptr()) })
     }
 
     /// Reverses the list in-place. Equivalent to the Python expression `l.reverse()`.
-    pub fn reverse(&self) -> PyResult<()> {
+    fn reverse(&self) -> PyResult<()> {
         err::error_on_minusone(self.py(), unsafe { ffi::PyList_Reverse(self.as_ptr()) })
     }
 
     /// Return a new tuple containing the contents of the list; equivalent to the Python expression `tuple(list)`.
     ///
     /// This method is equivalent to `self.as_sequence().to_tuple()` and faster than `PyTuple::new(py, this_list)`.
-    pub fn to_tuple(&self) -> &PyTuple {
-        unsafe { self.py().from_owned_ptr(ffi::PyList_AsTuple(self.as_ptr())) }
+    fn to_tuple(&self) -> Py2<'py, PyTuple> {
+        unsafe {
+            ffi::PyList_AsTuple(self.as_ptr())
+                .assume_owned(self.py())
+                .downcast_into_unchecked()
+        }
     }
 }
 
-index_impls!(PyList, "list", PyList::len, PyList::get_slice);
```

Comment:
Ah, yes, this is one of the sadder parts about the `Py2` APIs. Because index operators expect references, we just flat-out can't have this with `Py2`. I'd forgotten to point this out, but I also don't see a way around this.

(Given that we are more or less decided that we're going to move ahead with Py2 for 0.21, I think I will start building a migration guide entry now.)

Reply:
Yeah, that doesn't work. Most likely because those traits predate generic associated types. Do we have any idea how common it is to access `PyList` via `Index`?

Reply:
I suspect it's a popular convenience but I still don't see a way to let users keep it 😭

Reply:
Fair enough.